File: 2022/02.rs
type Input = Vec<(u32, u32)>;

fn get_input(input: &str) -> Input {
    input
        .trim()
        .split('\n')
        .map(|line| {
            (
                (line.as_bytes()[0] - b'A') as u32,
                (line.as_bytes()[2] - b'X') as u32,
            )
        })
        .collect()
}

fn part1(input: &Input) -> u32 {
    input.iter().map(|(a, b)| (4 + b - a) % 3 * 3 + b + 1).sum()
}

fn part2(input: &Input) -> u32 {
    input.iter().map(|(a, b)| (2 + a + b) % 3 + 1 + b * 3).sum()
}

aoc::main!(2022, 2);

File: 2022/02.py
def part1(puzzle: str):
    out = 0
    for line in puzzle.splitlines():
        a, b = line.split()
        a = ord(a) - ord("A")
        b = ord(b) - ord("X")
        w = (a - b) % 3
        if w == 0:
            out += 3 + b + 1
        elif w == 1:
            out += b + 1
        elif w == 2:
            out += 6 + b + 1
    return out


def part2(puzzle: str):
    out = 0
    for line in puzzle.splitlines():
        a, b = line.split()
        a = ord(a) - ord("A")
        b = ord(b) - ord("X")
        w = (a + b - 1) % 3
        if b == 0:
            out += w + 1
        elif b == 1:
            out += 3 + w + 1
        elif b == 2:
            out += 6 + w + 1
    return out


if __name__ == "__main__":
    from aoc import run

    run(2022, 2, part1, part2)
